File: dereplication_pilot/amr_and_coassembly_reports.py
#!/usr/bin/env python3
"""Two more reports built on the same dereplication clustering, no new compute:

1. AMR redundancy: how many raw RGI hits exist across all 296 assemblies vs.
   how many *distinct* AMR genes that collapses to once cluster membership
   is accounted for (i.e., the real, non-redundant AMR gene count).
2. Co-assembly-unique contribution: for the 19 participants with a co-assembly,
   how many of its contigs/genes fall in a cluster with NO single-sample member
   at all -- i.e., only recoverable by pooling reads across timepoints, not by
   any individual single-sample assembly. Cross-referenced against AMR hits to
   flag any resistance gene the co-assembly caught that no single sample would
   have.
"""
import csv
import glob
from collections import defaultdict

from validate_propagation import (
    PILOT_DIR,
    is_coassembly,
    load_gene_calls_rgi,
)


def load_membership(path):
    cluster_of = {}
    members_of = defaultdict(list)
    with open(path) as fh:
        reader = csv.DictReader(fh, delimiter="\t")
        for row in reader:
            cluster_of[row["member_id"]] = row["cluster_id"]
            members_of[row["cluster_id"]].append((row["member_id"], row["source_group"]))
    return cluster_of, members_of


def amr_redundancy_report(out_path):
    lines = ["=== AMR redundancy (RGI Best_Hit_ARO), all participants ===", ""]
    total_raw = 0
    total_distinct_clusters = 0
    per_participant = []

    for mfile in sorted(glob.glob(f"{PILOT_DIR}/p*/orfs/cluster_membership.tsv")):
        participant = mfile.split("/")[-3]
        cluster_of, _ = load_membership(mfile)
        groups = set()
        with open(mfile) as fh:
            reader = csv.DictReader(fh, delimiter="\t")
            for row in reader:
                groups.add(row["source_group"])

        raw_hits = 0
        amr_clusters = set()
        for g in groups:
            calls = load_gene_calls_rgi(g)
            raw_hits += len(calls)
            for gene_id in calls:
                cid = cluster_of.get(gene_id)
                if cid is not None:
                    amr_clusters.add(cid)

        total_raw += raw_hits
        total_distinct_clusters += len(amr_clusters)
        per_participant.append((participant, raw_hits, len(amr_clusters)))

    lines.append(f"total raw RGI hits (across all groups, all participants): {total_raw}")
    lines.append(f"total distinct AMR gene clusters (deduplicated): {total_distinct_clusters}")
    if total_raw:
        lines.append(
            f"redundancy: {100.0 * (total_raw - total_distinct_clusters) / total_raw:.1f}% "
            f"of raw AMR hits are duplicates of another hit already counted"
        )
    lines.append("")
    lines.append("per-participant (raw hits, distinct AMR clusters):")
    for p, raw, distinct in per_participant:
        lines.append(f"  {p}: {raw}, {distinct}")

    report = "\n".join(lines) + "\n"
    with open(out_path, "w") as fh:
        fh.write(report)
    print(report, end="")


def coassembly_unique_report(out_path):
    lines = ["=== co-assembly-unique contigs/genes (19 participants with a co-assembly) ===", ""]

    for tier in ("contigs", "orfs"):
        lines.append(f"--- tier: {tier} ---")
        for mfile in sorted(glob.glob(f"{PILOT_DIR}/p*/{tier}/cluster_membership.tsv")):
            participant = mfile.split("/")[-3]
            cluster_of, members_of = load_membership(mfile)

            coassembly_group = None
            for cid, members in members_of.items():
                for member_id, group in members:
                    if is_coassembly(group):
                        coassembly_group = group
                        break
                if coassembly_group:
                    break
            if not coassembly_group:
                continue  # one of the 6 no-co-assembly participants

            ca_total = 0
            ca_unique = 0
            unique_amr_hits = []
            rgi_calls = load_gene_calls_rgi(coassembly_group) if tier == "orfs" else {}

            for cid, members in members_of.items():
                groups_in_cluster = set(g for _, g in members)
                ca_members = [m for m, g in members if g == coassembly_group]
                if not ca_members:
                    continue
                ca_total += len(ca_members)
                if groups_in_cluster == {coassembly_group}:
                    ca_unique += len(ca_members)
                    if tier == "orfs":
                        for m in ca_members:
                            if m in rgi_calls:
                                unique_amr_hits.append((m, rgi_calls[m]))

            pct = 100.0 * ca_unique / ca_total if ca_total else 0.0
            lines.append(
                f"  {participant} ({coassembly_group}): {ca_unique}/{ca_total} "
                f"({pct:.1f}%) co-assembly-only, no single-sample match"
            )
            if unique_amr_hits:
                for gene_id, call in unique_amr_hits:
                    lines.append(f"    AMR hit only found via co-assembly: {gene_id} -> {call}")
        lines.append("")

    report = "\n".join(lines) + "\n"
    with open(out_path, "w") as fh:
        fh.write(report)
    print(report, end="")


if __name__ == "__main__":
    import sys

    amr_redundancy_report(f"{PILOT_DIR}/amr_redundancy.txt")
    print()
    coassembly_unique_report(f"{PILOT_DIR}/coassembly_unique.txt")

File: dereplication_pilot/R/amr_and_coassembly_reports.R
# R port of amr_and_coassembly_reports.py:
#  1. AMR redundancy: raw RGI hits vs. distinct AMR gene clusters, per participant.
#  2. Co-assembly-unique contribution: contigs/genes in a cluster with NO single-sample
#     member at all, per participant/tier, cross-referenced against AMR hits.
# Run from dereplication_pilot/.

source("R/utils.R")

amr_redundancy_report <- function() {
  participants <- all_participants()
  per_participant <- list()

  for (participant in participants) {
    membership <- load_cluster_membership(participant, "orfs")
    groups <- unique(membership$source_group)
    calls <- purrr::map_dfr(groups, load_gene_calls_rgi)

    joined <- membership %>% inner_join(calls, by = c("member_id" = "gene_id"))
    per_participant[[participant]] <- tibble(
      participant = participant,
      raw_hits = nrow(joined),
      distinct_clusters = n_distinct(joined$cluster_id)
    )
  }

  result <- bind_rows(per_participant)
  write_tsv(result %>% select(participant, raw_hits, distinct_clusters),
            file.path(PILOT_DIR, "R_report_amr_redundancy.tsv"))

  total_raw <- sum(result$raw_hits)
  total_distinct <- sum(result$distinct_clusters)
  cat(sprintf("total raw RGI hits: %d\ntotal distinct AMR gene clusters: %d\nredundancy: %.1f%%\n",
              total_raw, total_distinct, 100 * (total_raw - total_distinct) / total_raw))
  result
}

coassembly_unique_report <- function() {
  participants <- all_participants()
  unique_rows <- list()
  amr_unique_rows <- list()

  for (participant in participants) {
    for (tier in c("contigs", "orfs")) {
      membership <- load_cluster_membership(participant, tier)
      coassembly_group <- membership$source_group[is_coassembly(membership$source_group)][1]
      if (is.na(coassembly_group)) next  # one of the 6 no-co-assembly participants

      cluster_groups <- membership %>%
        distinct(cluster_id, source_group) %>%
        group_by(cluster_id) %>%
        summarise(n_distinct_groups = n_distinct(source_group),
                  has_coassembly = any(is_coassembly(source_group)), .groups = "drop")

      ca_members <- membership %>% filter(source_group == coassembly_group)
      ca_total <- nrow(ca_members)

      unique_cluster_ids <- cluster_groups %>%
        filter(has_coassembly, n_distinct_groups == 1) %>% pull(cluster_id)
      ca_unique_members <- ca_members %>% filter(cluster_id %in% unique_cluster_ids)
      ca_unique <- nrow(ca_unique_members)

      unique_rows[[paste(participant, tier)]] <- tibble(
        participant = participant, coassembly_group = coassembly_group, tier = tier,
        unique_count = ca_unique, total_count = ca_total,
        unique_pct = round(100 * ca_unique / ca_total, 1)
      )

      if (tier == "orfs" && ca_unique > 0) {
        calls <- load_gene_calls_rgi(coassembly_group)
        hits <- ca_unique_members %>% inner_join(calls, by = c("member_id" = "gene_id"))
        if (nrow(hits) > 0) {
          amr_unique_rows[[paste(participant, tier)]] <- hits %>%
            transmute(participant, gene_id = member_id, aro_call)
        }
      }
    }
  }

  list(
    unique = bind_rows(unique_rows),
    amr_unique = bind_rows(amr_unique_rows)
  )
}

if (TRUE) {  # always run main logic on source() or Rscript alike -- sys.nframe()==0 broke under nested source()
  amr_redundancy_report()
  result <- coassembly_unique_report()
  write_tsv(result$unique, file.path(PILOT_DIR, "R_report_coassembly_unique.tsv"))
  write_tsv(result$amr_unique, file.path(PILOT_DIR, "R_report_amr_coassembly_unique.tsv"))
  cat(sprintf("co-assembly-unique: %d rows, AMR-unique hits: %d\n",
              nrow(result$unique), nrow(result$amr_unique)))
}
